README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# modeltime.gluonts <img src='man/figures/logo-modeltime-gluonts.png' align="right" height="138" />

<!-- badges: start -->

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/modeltime.gluonts)](https://cran.r-project.org/package=modeltime)
![](http://cranlogs.r-pkg.org/badges/modeltime.gluonts?color=brightgreen)
![](http://cranlogs.r-pkg.org/badges/grand-total/modeltime.gluonts?color=brightgreen)
[![Codecov test
coverage](https://codecov.io/gh/business-science/modeltime.gluonts/branch/master/graph/badge.svg)](https://codecov.io/gh/business-science/modeltime.gluonts?branch=master)
[![R-CMD-check](https://github.com/business-science/modeltime.gluonts/workflows/R-CMD-check/badge.svg)](https://github.com/business-science/modeltime.gluonts/actions)
<!-- badges: end -->

> GluonTS Deep Learning in R.

## GluonTS Deep Learning in R

Modeltime GluonTS integrates the **Python GluonTS Deep Learning
Library**, making it easy to develop forecasts using Deep Learning for
those that are comfortable with the [Modeltime Forecasting
Workflow](https://business-science.github.io/modeltime/).

<img src="man/figures/m4_hourly_forecast.jpg" width="90%" style="display: block; margin: auto;" />

## Installation Requirements

**Important: This package is being maintained on GitHub (not CRAN).
Please install the GitHub version, which is updated with the latest
features:**

    # Install GitHub Version 
    remotes::install_github("business-science/modeltime.gluonts")

    # Install Python Dependencies
    modeltime.gluonts::install_gluonts()

For more detailed installation instructions and troubleshooting
guidance, visit our [Installation
Guide](https://business-science.github.io/modeltime.gluonts/articles/managing-envs.html).

## Make Your First DeepAR Model

Make your first `deep_ar()` model, which connects to the GluonTS
`DeepAREstimator()`. For a more detailed walkthough, visit our [Getting
Started
Guide](https://business-science.github.io/modeltime.gluonts/articles/getting-started.html).

``` r
library(modeltime.gluonts)
library(tidymodels)
library(tidyverse)

# Fit a GluonTS DeepAR Model
model_fit_deepar <- deep_ar(
    id                    = "id",
    freq                  = "M",
    prediction_length     = 24,
    lookback_length       = 48,
    epochs                = 5
) %>%
    set_engine("gluonts_deepar") %>%
    fit(value ~ ., training(m750_splits))

# Forecast with 95% Confidence Interval
modeltime_table(
    model_fit_deepar
) %>%
    modeltime_calibrate(new_data = testing(m750_splits)) %>%
    modeltime_forecast(
        new_data      = testing(m750_splits),
        actual_data   = m750,
        conf_interval = 0.95
    ) %>%
    plot_modeltime_forecast(.interactive = FALSE)
```

<img src="man/figures/deepar_example_1.png" width="90%" style="display: block; margin: auto;" />

## Meet the modeltime ecosystem

> Learn a growing ecosystem of forecasting packages

<div class="figure" style="text-align: center">

<img src="man/figures/modeltime_ecosystem.jpg" alt="The modeltime ecosystem is growing" width="100%" />
<p class="caption">
The modeltime ecosystem is growing
</p>

</div>

Modeltime is part of a **growing ecosystem** of Modeltime forecasting
packages.

-   [Modeltime (Machine
    Learning)](https://business-science.github.io/modeltime/)

-   [Modeltime H2O
    (AutoML)](https://business-science.github.io/modeltime.h2o/)

-   [Modeltime GluonTS (Deep
    Learning)](https://business-science.github.io/modeltime.gluonts/)

-   [Modeltime Ensemble (Blending
    Forecasts)](https://business-science.github.io/modeltime.ensemble/)

-   [Modeltime Resample
    (Backtesting)](https://business-science.github.io/modeltime.resample/)

-   [Timetk (Feature Engineering, Data Wrangling, Time Series
    Visualization)](https://business-science.github.io/timetk/)

## Take the High-Performance Forecasting Course

> Become the forecasting expert for your organization

<a href="https://university.business-science.io/p/ds4b-203-r-high-performance-time-series-forecasting/" target="_blank"><img src="https://www.filepicker.io/api/file/bKyqVAi5Qi64sS05QYLk" alt="High-Performance Time Series Forecasting Course" width="100%" style="box-shadow: 0 0 5px 2px rgba(0, 0, 0, .5);"/></a>

[*High-Performance Time Series
Course*](https://university.business-science.io/p/ds4b-203-r-high-performance-time-series-forecasting/)

### Time Series is Changing

Time series is changing. **Businesses now need 10,000+ time series
forecasts every day.** This is what I call a *High-Performance Time
Series Forecasting System (HPTSF)* - Accurate, Robust, and Scalable
Forecasting.

**High-Performance Forecasting Systems will save companies by improving
accuracy and scalability.** Imagine what will happen to your career if
you can provide your organization a “High-Performance Time Series
Forecasting System” (HPTSF System).

### How to Learn High-Performance Time Series Forecasting

I teach how to build a HPTFS System in my [**High-Performance Time
Series Forecasting
Course**](https://university.business-science.io/p/ds4b-203-r-high-performance-time-series-forecasting).
You will learn:

-   **Time Series Machine Learning** (cutting-edge) with `Modeltime` -
    30+ Models (Prophet, ARIMA, XGBoost, Random Forest, & many more)
-   **Deep Learning** with `GluonTS` (Competition Winners)
-   **Time Series Preprocessing**, Noise Reduction, & Anomaly Detection
-   **Feature engineering** using lagged variables & external regressors
-   **Hyperparameter Tuning**
-   **Time series cross-validation**
-   **Ensembling** Multiple Machine Learning & Univariate Modeling
    Techniques (Competition Winner)
-   **Scalable Forecasting** - Forecast 1000+ time series in parallel
-   and more.

<p class="text-center" style="font-size:24px;">
Become the Time Series Expert for your organization.
</p>
<br>
<p class="text-center" style="font-size:30px;">
<a href="https://university.business-science.io/p/ds4b-203-r-high-performance-time-series-forecasting">Take
the High-Performance Time Series Forecasting Course</a>
</p>


NEWS.md

# modeltime.gluont 0.3.1

## Installation Support

### Windows Conflicting Dependencies

Improved support for conflicting package dependencies on Windows Operating Systems. Solution is to separate the installation process into two stages, which happens inside of `install_gluonts()`. #32

```
pytorch-lightning 1.3.8 depends on numpy>=1.17.2
mxnet 1.7.0.post1 depends on numpy<1.17.0 and >=1.8.2
```

### New Uninstall Function

Users can now `uninstall_gluonts()`.



# modeltime.gluonts 0.3.0

## Support for GluonTS 0.8.0 and Pytorch Backend:

Modeltime GluonTS now support `gluonts 0.8.0`. Simply run `install_gluonts()` to upgrade. The upgraded support makes `modeltime.gluonts` incompatible with earlier versions of GluonTS (e.g. `gluonts 0.6.3`). The solution is to upgrade to `gluonts 0.8.0`, which requires:

- `gluonts==0.8.0`

- `mxnet~=1.7`

Additionally, GluonTS 0.8.0 now supports __pytorch as a backend.__ Use `install_gluonts(include_pytorch = TRUE)` to simplify installation of the PyTorch backend. Pytorch backend requirements:

- `torch~=1.6.0`
            
- `pytorch-lightning~=1.1`

## New Algorithms

### Pytorch  DeepAR 

A new engine has been added to `deep_ar()` that enables the Pytorch backend using `set_engine("torch")`. This requires the Python packages `pytorch` and `pytorch-lightning`. Use `install_gluonts(include_pytorch = TRUE)` to simplify installation.

### GP Forecaster Algorithm

A new function, `gp_forecaster()`, integrates the Gaussian Process Estimator from GluonTS. 

### Deep State Algorithm

A new function, `deep_state()`, integrates the Deep State Estimator from GluonTS. 

## Tutorials

- We've updated the [Installation Guide](https://business-science.github.io/modeltime.gluonts/articles/managing-envs.html). This includes revised requirements for installation, upgrading to `modeltime.gluonts` >= 0.3.0, troubleshooting installation, python environment requirements, and custom python environments. 

- We've updated the [Getting Started Guide](https://business-science.github.io/modeltime.gluonts/articles/getting-started.html) to go through a DeepAR example.

- We've update the [GPU Setup Instructions](https://business-science.github.io/modeltime.gluonts/articles/using-gpus.html) to cover Modeltime >=0.3.0.

## Improvements

- `install_gluonts()`: Gains two new parameters to help with upgrading:
    1. `fresh_install`: If TRUE, will remove prior installations of the r-glounts conda environment to setup for a fresh installation. This can be useful if errors appear during upgrades. Default: FALSE.
    2. `include_pytorch`: If TRUE, will install torch. Needed for Torch implementation of deep_ar(). Default: FALSE.

## Breaking Changes

- GluonTS <= 0.8.0. The `modeltime.gluonts` package version >= 0.2.2.9000 is not compatible with `gluonts` < 0.8.0. To fix, simply upgrade to `gluonts` 0.8.0. 

# modeltime.gluonts 0.2.2

### Dials Params

- NBEATS Models: Adding Dials helpers #14

# modeltime.gluonts 0.2.1

Improvements made to connect with the GluonTS Python Environment on Startup. 

# modeltime.gluonts 0.2.0

### New Vignettes

- [Managing GluonTS Environments](https://business-science.github.io/modeltime.gluonts/articles/managing-envs.html)

- [Using GPUs with GluonTS](https://business-science.github.io/modeltime.gluonts/articles/using-gpus.html)

### New Features

- __Internal Scaling by Group:__ After significant testing it appears that some data sets return better results when the data is scaled by time series "id" (group). To help facilitate this, a new option is available scale by id: `scale = TRUE`.

- __Custom Python Environments:__ Provide an option for setting a Custom Python Environment by supplying a `GLUONTS_PYTHON` environment variable. Before running `library(modeltime.gluonts)` use `Sys.setenv(GLUONTS_PYTHON = 'path/to/python')` to set the path of your python executable in a Conda or Virtual Environment that has 'gluonts', 'mxnet', 'numpy', 'pandas' and 'pathlib' available as dependencies.

### Fixes & Improvements

* __GluonTS 0.6.3 Upgrade:__ `install_gluonts()` now uses `gluonts==0.6.3`. This upgrade improves forecast accuracy.
* CRAN Comment - Add `SystemRequirements`: GluonTS.
* CRAN Comment - Fix `.onLoad` message to provide options for configuring the Python Environment.

# modeltime.gluonts 0.1.0

* __Models:__ Initial Release incorporates 2 GluonTS Algorithms:

    - `deep_ar()`: Integrates GluonTS DeepAREstimator
    - `nbeats()`: Integrates N-BEATS & N-BEATS Ensemble Estimators
    
* __New Vignette:__ [Getting Started](https://business-science.github.io/modeltime.gluonts/articles/getting-started.html)

* __Website:__ [Modeltime GluonTS](https://business-science.github.io/modeltime.gluonts/)


R/core-install.R
#' Install / Uninstall GluonTS
#'
#' @description
#' `install_gluonts()`: Installs `GluonTS` Probabilisitic Deep Learning Time Series Forecasting Software
#' using `reticulate::py_install()`.
#'
#' - A `Python` Environment will be created named `r-gluonts`.
#' - When loaded with `library(modeltime.gluonts)`, the `modeltime.gluonts` R package
#'   will connect to the `r-gluonts` Python environment by default. See "Details" for
#'   connecting to custom python environments.
#' - If `fresh_install`, will remove any prior installations of the "r-gluonts" python environment
#' - If `include_pytorch`, will install additional dependencies needed for the optional
#'   pytorch backend that is available in some algorithms.
#'
#' `uninstall_gluonts()`: Will remove the "r-gluonts" python environment and python packages
#'
#' @param include_pytorch If `TRUE`, will install `torch`. Needed for Torch implementation
#'   of `deep_ar()`. Default: `FALSE`.
#' @param fresh_install If `TRUE`, will remove prior installations of the `r-glounts`
#'   conda environment to setup for a fresh installation. This can be useful if
#'   errors appear during upgrades. Default: `FALSE`.
#'
#' @details
#'
#' __Options for Connecting to Python__
#'
#' - __Recommended__ _Use Pre-Configured Python Environment:_ Use `install_gluonts()` to
#'    install GluonTS Python Libraries into a conda environment named 'r-gluonts'.
#' - __Advanced__ _Use a Custom Python Environment:_ Before running `library(modeltime.gluonts)`,
#'    use `Sys.setenv(GLUONTS_PYTHON = 'path/to/python')` to set the path of your
#'    python executable in an environment that has 'gluonts', 'mxnet', 'numpy', 'pandas',
#'    and 'pathlib' available as dependencies.
#'
#' __Package Manager Support (Python Environment)__
#'
#' - __Conda Environments:__ Currently, `install_gluonts()` supports Conda and Miniconda Environments.
#'
#' - __Virtual Environments:__ are not currently supported with the default installation method, `install_gluonts()`.
#'    However, you can connect to virtual environment that you have created using
#'     `Sys.setenv(GLUONTS_PYTHON = 'path/to/python')` prior to running `library(modeltime.ensemble)`.
#'
#' @examples
#' \dontrun{
#' install_gluonts()
#' }
#'
#'
#' @export
install_gluonts <- function(
    fresh_install = FALSE,
    include_pytorch = FALSE
) {

    # Check for Anaconda
    if (!check_conda()) {
        return()
    }

    # REMOVE PREVIOUS ENV
    if (fresh_install) {
        cli::cli_alert_info("Removing conda env `r-gluonts` to setup for fresh install...")
        reticulate::conda_remove("r-gluonts")
    }


    # PYTHON SPEC
    python_version <- "3.7.1"

    # GLUONTS INSTALLATION

    message("\n")
    cli::cli_alert_info("Installing gluonts dependencies...")
    message("\n")

    default_pkgs <- c(
        "mxnet~=1.7",
        "gluonts==0.8.0",
        "numpy",
        "pandas==1.0.5",
        "pathlib==1.0.1",
        "ujson==4.0.2",
        "brotlipy"
    )


    reticulate::py_install(
        packages       = default_pkgs,
        envname        = "r-gluonts",
        method         = "conda",
        conda          = "auto",
        python_version = python_version,
        pip            = TRUE
    )


    # TORCH INSTALLATION
    if (include_pytorch) {

        message("\n")
        cli::cli_alert_info("Installing torch dependencies...")
        message("\n")

        torch_pkgs <- c(
            "torch~=1.6",
            "pytorch-lightning~=1.1"
        )

        reticulate::py_install(
            packages       = torch_pkgs,
            envname        = "r-gluonts",
            method         = "conda",
            conda          = "auto",
            python_version = "3.7.1",
            pip            = TRUE
        )

    }

    # PROCESS CHECKS
    env_exists      <- !is.null(detect_default_gluonts_env())
    gluonts_failure <- FALSE

    message("\n")
    if (env_exists) {
        cli::cli_alert_success("The {.field r-gluonts} conda environment has been created.")
    } else {
        cli::cli_alert_danger("The {.field r-gluonts} conda environment could not be created.")
        gluonts_failure <- TRUE
    }

    # default_pkgs_exist <- check_gluonts_dependencies()
    # if (default_pkgs_exist) {
    #     cli::cli_alert_success("Installing gluonts dependencies... ...Success.")
    # } else {
    #     cli::cli_alert_danger("Installing gluonts dependencies... ...Failed. One or more of the following packages are not available: gluonts, mxnet, numpy, pandas, pathlib")
    #     gluonts_failure <- TRUE
    # }
    #
    # pytorch_failure <- FALSE
    # if (include_pytorch) {
    #     pytorch_exists     <- check_pytorch_dependencies()
    #
    #     if (pytorch_exists) {
    #         cli::cli_alert_success("Installing torch dependencies... ...Success.")
    #     } else {
    #         cli::cli_alert_danger("Installing torch dependencies... ...Failed. One or more of the following packages are not available: torch, pytorch_lightning")
    #         pytorch_failure <- TRUE
    #     }
    #
    #
    # }

    if (env_exists) {
        cli::cli_alert_info("Please restart your R Session and run {.code library(modeltime.gluonts)} to activate the {.field r-gluonts} environment.")
    } else {
        cli::cli_alert_info("For installation failure reports, please copy the python error message(s). Search your error(s) using Google. If none are found, create new issues here: https://github.com/business-science/modeltime.gluonts/issues")
    }

}


#' @export
#' @rdname install_gluonts
uninstall_gluonts <- function() {
    cli::cli_alert_info("Removing conda env `r-gluonts`...")
    reticulate::conda_remove("r-gluonts")
    message("\n")
    cli::cli_alert_success("The `r-gluonts` env has been removed.")
}


check_conda <- function() {

    conda_list_nrow <- nrow(reticulate::conda_list())

    if (is.null(conda_list_nrow) || conda_list_nrow == 0L) {
        # No conda
        message("Could not detect Conda or Miniconda Package Managers, one of which is required for 'install_gluonts()'. \nAvailable options:\n",
                " - [Preferred] You can install Miniconda (light-weight) using 'reticulate::install_miniconda()'. \n",
                " - Or, you can install the full Aniconda distribution (1000+ packages) using 'reticulate::conda_install()'. \n\n",
                "Then use 'install_gluonts()' to set up the GluonTS python environment.")
        conda_found <- FALSE
    } else {
        conda_found <- TRUE
    }

    return(conda_found)
}

